### ISSUE TITLE
[python-package] Where is the code for the 'quantile' objective, and how do I pass a custom objective to LGBMRegressor?

### ISSUE THREAD

**Issue by qinhaohcen** [microsoft/LightGBM#5524]
## Description

I want to change quantiles' loss function. Although I can make objective='quantile' , but I can't see a quantile in LGBMRegressor's code. Can you give me some advice? Thank you!

**Comment by jameslamb** [comment]
Thanks for using LightGBM! I'm confused a bit by your question.

Which of these are you asking?

* "how do I provide a custom objective function for quantile regression with `LGBMRegressor` in the Python package?"
* "where can I find the code for the quantile objective function, so that I understand how it works?"
* "what parameters can I use to change the behavior of LightGBM's quantile regression?"

Can you please tell me which of those you're asking?

**Comment by qinhaohcen** [comment]
Thank you for your patient reply！ I would like to ask for advice on this three points you mentioned.

"how do I provide a custom objective function for quantile regression with LGBMRegressor in the Python package?"
"where can I find the code for the quantile objective function, so that I understand how it works?"
"what parameters can I use to change the behavior of LightGBM's quantile regression?"

**Comment by qinhaohcen** [comment]
> Thanks for using LightGBM! I'm confused a bit by your question.
> 
> Which of these are you asking?
> 
> * "how do I provide a custom objective function for quantile regression with `LGBMRegressor` in the Python package?"
> * "where can I find the code for the quantile objective function, so that I understand how it works?"
> * "what parameters can I use to change the behavior of LightGBM's quantile regression?"
> 
> Can you please tell me which of those you're asking?

This is my code. When I move the mouse to 'quantite' and press Ctrl+B, I will be prompted to Can't find declaration to go to. But I really want to find out how the quantile method works in `LGBMRegenerator`. And I hope to improve the loss function pinball loss of quantile.
`        lgb_model_q05 = LGBMRegressor(objective='quantile',
                                      metric='quantile',
                                      alpha=0.05,
                                      learning_rate=0.1,
                                      max_depth=7,
                                      n_estimators=100)`

**Comment by jameslamb** [comment]
**"how do I provide a custom objective function for quantile regression with LGBMRegressor in the Python package?"**

See the `Note` on custom objectives at the bottom of https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html#lightgbm.LGBMRegressor for information on how to format a custom objective function for `LGBMRegressor`.

Custom objective functions can be passed to the `objective` parameter.

<hr>

**"where can I find the code for the quantile objective function, so that I understand how it works?"**

the built-in `quantile` loss is defined here

https://github.com/microsoft/LightGBM/blob/c5391c9701122a0779937f380a4108b5276e14fa/src/objective/regression_objective.hpp#L481

<hr>

**"what parameters can I use to change the behavior of LightGBM's quantile regression?"**

The only quantile-specific parameter is `alpha` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#alpha).

See https://lightgbm.readthedocs.io/en/latest/Parameters.html for descriptions of the other parameters that control LightGBM's behavior.

**Comment by qinhaohcen** [comment]
Thank you very much！Now I know why I can't find it.

**Comment by qinhaohcen** [comment]
> 
Why is hessian directly assigned to 1.0? Is there any theoretical basis? Can you provide a PDF materials for reference？
``

**Comment by qinhaohcen** [comment]
> **"how do I provide a custom objective function for quantile regression with LGBMRegressor in the Python package?"**
> 
> See the `Note` on custom objectives at the bottom of https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html#lightgbm.LGBMRegressor for information on how to format a custom objective function for `LGBMRegressor`.
> 
> Custom objective functions can be passed to the `objective` parameter.
> 
> **"where can I find the code for the quantile objective function, so that I understand how it works?"**
> 
> the built-in `quantile` loss is defined here
> 
> https://github.com/microsoft/LightGBM/blob/c5391c9701122a0779937f380a4108b5276e14fa/src/objective/regression_objective.hpp#L481
> 
> **"what parameters can I use to change the behavior of LightGBM's quantile regression?"**
> 
> The only quantile-specific parameter is `alpha` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#alpha).
> 
> See https://lightgbm.readthedocs.io/en/latest/Parameters.html for descriptions of the other parameters that control LightGBM's behavior.

If you use pinball loss for quantile，gradient（First derivative）should be 1-alpha or alpha. And hessian should be 0. 

Did you consider the combination of hessian and lamda and assign hessian as 1？

Sincerely look forward to your answer!

**Comment by qinhaohcen** [comment]
> **"how do I provide a custom objective function for quantile regression with LGBMRegressor in the Python package?"**
> 
> See the `Note` on custom objectives at the bottom of https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html#lightgbm.LGBMRegressor for information on how to format a custom objective function for `LGBMRegressor`.
> 
> Custom objective functions can be passed to the `objective` parameter.
> 
> **"where can I find the code for the quantile objective function, so that I understand how it works?"**
> 
> the built-in `quantile` loss is defined here
> 
> https://github.com/microsoft/LightGBM/blob/c5391c9701122a0779937f380a4108b5276e14fa/src/objective/regression_objective.hpp#L481
> 
> **"what parameters can I use to change the behavior of LightGBM's quantile regression?"**
> 
> The only quantile-specific parameter is `alpha` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#alpha).
> 
> See https://lightgbm.readthedocs.io/en/latest/Parameters.html for descriptions of the other parameters that control LightGBM's behavior.

If you use pinball loss for quantile，gradient（First derivative）should be 1-alpha or alpha. And hessian should be 0. 

Did you consider the combination of hessian and lamda and assign hessian as 1？

Sincerely look forward to your answer!

**Comment by jmoralez** [comment]
I believe it's an implementation detail where the sum of the hessians is used as a proxy to the number of samples in a node, but hopefully @shiyu1994 or @guolinke can confirm.
